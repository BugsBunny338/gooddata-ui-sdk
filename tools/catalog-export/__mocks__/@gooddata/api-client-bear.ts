// (C) 2007-2018 GoodData Corporation
import catalogItemsResponse from "./loadItems.json";
import getObjectsResponse from "./getObjects.json";
import getObjectsByQueryResponse from "./getObjectsByQuery.json";
import getVisualizationsResponse from "./getVisualizations.json";

export default {
    catalogue: {
        loadItems() {
            return new Promise((resolve) => resolve(catalogItemsResponse));
        },
    },
    md: {
        getObjectsByQuery() {
            return new Promise((resolve) => resolve(getObjectsByQueryResponse));
        },
        getObjects() {
            return new Promise((resolve) => resolve(getObjectsResponse));
        },
        getVisualizations() {
            return new Promise((resolve) => resolve(getVisualizationsResponse));
        },
    },
};
