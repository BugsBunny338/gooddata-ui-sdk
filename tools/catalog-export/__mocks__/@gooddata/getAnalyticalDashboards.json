[
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/2e5a40d590f955c88251a8aead5267bd",
        "created": "2017-01-26 17:53:13",
        "deprecated": "0",
        "identifier": "afMA17GSbk31",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "KPIs",
        "updated": "2020-11-18 22:14:11",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/2663"
    },
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "created": "2017-05-03 15:03:50",
        "deprecated": "0",
        "identifier": "abBJlnxrfEWH",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "Store management KPIs",
        "updated": "2018-01-23 04:43:37",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/6853"
    },
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "created": "2018-11-12 11:29:01",
        "deprecated": "0",
        "identifier": "aahlrtUEgKkq",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "Untitled",
        "updated": "2018-11-12 11:29:02",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/8867"
    }
]
