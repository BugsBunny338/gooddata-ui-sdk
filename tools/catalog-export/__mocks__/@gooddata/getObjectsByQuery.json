[
    {
        "dataSet": {
            "content": {
                "urn": "urn:gooddata:date",
                "mode": "",
                "facts": [],
                "titleSuffix": "Timeline",
                "dataLoadingColumns": [],
                "attributes": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/439",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/440",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/441",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/442",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/443",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/444",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/445",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/446",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/447",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/448",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/449",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/450",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/451",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/452",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/453",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/454",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/455",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/456"
                ],
                "identifierPrefix": "timeline"
            },
            "links": {
                "dataUploads": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/data/uploads/499"
            },
            "meta": {
                "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
                "uri": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/499",
                "tags": "",
                "created": "2010-08-02 08:13:42",
                "identifier": "timeline.dataset.dt",
                "deprecated": "0",
                "summary": "DataSet Date",
                "isProduction": 1,
                "title": "Date (Timeline)",
                "category": "dataSet",
                "updated": "2018-05-15 16:18:26",
                "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
            }
        }
    },
    {
        "dataSet": {
            "content": {
                "mode": "SLI",
                "facts": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/706",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/707"
                ],
                "customUploadTimestamp": 1566890302945695,
                "dataLoadingColumns": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/62",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/64",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/132",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/134",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/17",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/19",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/21",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/23",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/25",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/27",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/29",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/31",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/33",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/35",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/37"
                ],
                "attributes": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/679",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/680",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/682"
                ]
            },
            "links": {
                "dataUploads": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/data/uploads/165"
            },
            "meta": {
                "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
                "uri": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/165",
                "tags": "",
                "created": "2018-04-27 11:13:04",
                "identifier": "dataset.salesdetailfact",
                "deprecated": "0",
                "summary": "",
                "isProduction": 1,
                "title": "Sales Detail Fact",
                "category": "dataSet",
                "updated": "2019-08-27 07:19:45",
                "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
            }
        }
    }
]
