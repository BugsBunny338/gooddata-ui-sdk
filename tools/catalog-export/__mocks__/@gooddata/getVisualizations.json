[
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/962",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "abhJpedgcfU2",
        "summary": "",
        "isProduction": 1,
        "title": "Labor Costs vs Scheduled Costs",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    },
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/964",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "acPWDTL2bJeX",
        "summary": "",
        "isProduction": 1,
        "title": "Total Sales by City",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    },
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/965",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "aaBN5UG3dXu4",
        "summary": "",
        "isProduction": 1,
        "title": "$ Total Sales by Location",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    }
]
