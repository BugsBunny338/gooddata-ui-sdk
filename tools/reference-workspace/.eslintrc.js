// (C) 2020 GoodData Corporation
module.exports = {
    extends: ["../../.eslintrc.js"],
    parserOptions: { tsconfigRootDir: __dirname, project: "./tsconfig.json" },
};
