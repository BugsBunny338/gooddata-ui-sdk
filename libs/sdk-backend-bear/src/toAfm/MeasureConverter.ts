// (C) 2007-2018 GoodData Corporation
import {
    IArithmeticMeasureDefinition,
    IMeasure,
    IMeasureDefinition,
    IMeasureDefinitionType,
    IPoPMeasureDefinition,
    IPreviousPeriodMeasureDefinition,
    isArithmeticMeasureDefinition,
    isMeasureDefinition,
    isPoPMeasureDefinition,
    isPreviousPeriodMeasureDefinition,
} from "@gooddata/sdk-model";
import { GdcExecuteAFM } from "@gooddata/gd-bear-model";
import { convertMeasureFilter } from "./FilterConverter";
import compact = require("lodash/compact");
import get = require("lodash/get");

export function convertMeasure(measure: IMeasure): GdcExecuteAFM.IMeasure {
    const {
        measure: { definition },
    } = measure;

    const convertedDefinition = convertMeasureDefinition(definition);

    const format = getFormat(measure);
    const formatProp = format ? { format } : {};

    const alias = measure.measure.alias ? measure.measure.alias : measure.measure.title;
    const aliasProp = alias ? { alias } : {};

    return {
        localIdentifier: measure.measure.localIdentifier,
        definition: convertedDefinition,
        ...aliasProp,
        ...formatProp,
    };
}

function convertMeasureDefinition(definition: IMeasureDefinitionType): GdcExecuteAFM.MeasureDefinition {
    if (isMeasureDefinition(definition)) {
        return convertSimpleMeasureDefinition(definition);
    } else if (isPoPMeasureDefinition(definition)) {
        return convertPopMeasureDefinition(definition);
    } else if (isPreviousPeriodMeasureDefinition(definition)) {
        return convertPreviousPeriodMeasureDefinition(definition);
    } else if (isArithmeticMeasureDefinition(definition)) {
        return convertArithmeticMeasureDefinition(definition);
    } else {
        throw Error("The measure definition is not supported: " + JSON.stringify(definition));
    }
}

function convertSimpleMeasureDefinition(
    definition: IMeasureDefinition,
): GdcExecuteAFM.ISimpleMeasureDefinition {
    const { measureDefinition } = definition;

    const filters: GdcExecuteAFM.FilterItem[] = measureDefinition.filters
        ? compact(measureDefinition.filters.map(convertMeasureFilter))
        : [];
    const filtersProp = filters.length ? { filters } : {};

    const aggregation = measureDefinition.aggregation;
    const aggregationProp = aggregation ? { aggregation } : {};

    const computeRatio = measureDefinition.computeRatio;
    const computeRatioProp = computeRatio ? { computeRatio } : {};

    return {
        measure: {
            item: measureDefinition.item,
            ...filtersProp,
            ...aggregationProp,
            ...computeRatioProp,
        },
    };
}

function convertPopMeasureDefinition(definition: IPoPMeasureDefinition): GdcExecuteAFM.IPopMeasureDefinition {
    const { popMeasureDefinition } = definition;
    return {
        popMeasure: {
            measureIdentifier: popMeasureDefinition.measureIdentifier,
            popAttribute: popMeasureDefinition.popAttribute,
        },
    };
}

function convertPreviousPeriodMeasureDefinition(
    definition: IPreviousPeriodMeasureDefinition,
): GdcExecuteAFM.IPreviousPeriodMeasureDefinition {
    const { previousPeriodMeasure } = definition;
    return {
        previousPeriodMeasure: {
            measureIdentifier: previousPeriodMeasure.measureIdentifier,
            dateDataSets: previousPeriodMeasure.dateDataSets.map(dateDataSet => ({
                dataSet: dateDataSet.dataSet,
                periodsAgo: dateDataSet.periodsAgo,
            })),
        },
    };
}

function convertArithmeticMeasureDefinition(
    definition: IArithmeticMeasureDefinition,
): GdcExecuteAFM.IArithmeticMeasureDefinition {
    const { arithmeticMeasure } = definition;
    return {
        arithmeticMeasure: {
            measureIdentifiers: arithmeticMeasure.measureIdentifiers.slice(),
            operator: arithmeticMeasure.operator,
        },
    };
}

function getFormat(measure: IMeasure): string | undefined {
    const {
        measure: { definition },
    } = measure;
    const measureFormat = get(measure.measure, "format");

    if (isArithmeticMeasureDefinition(definition)) {
        if (definition.arithmeticMeasure.operator === "change") {
            return "#,##0.00%";
        }
    }

    const predefinedFormat = isMeasureDefinition(definition) ? getPredefinedFormat(definition) : undefined;

    return predefinedFormat || measureFormat;
}

function getPredefinedFormat(definition: IMeasureDefinition): string | null {
    const { measureDefinition } = definition;
    // should we prefer format defined on measure? If so, fix computeRatio format in AD
    return measureDefinition.computeRatio
        ? "#,##0.00%"
        : measureDefinition.aggregation === "count"
        ? "#,##0"
        : null;
}
