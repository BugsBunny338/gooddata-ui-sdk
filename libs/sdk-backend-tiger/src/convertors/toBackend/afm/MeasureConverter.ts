// (C) 2007-2020 GoodData Corporation
import {
    IArithmeticMeasureDefinition,
    IMeasure,
    IMeasureDefinition,
    IMeasureDefinitionType,
    IPoPMeasureDefinition,
    IPreviousPeriodMeasureDefinition,
    isArithmeticMeasureDefinition,
    isMeasureDefinition,
    isPoPMeasureDefinition,
    isPreviousPeriodMeasureDefinition,
    MeasureAggregation,
} from "@gooddata/sdk-model";
import { ExecuteAFM } from "@gooddata/gd-tiger-client";
import { convertVisualizationObjectFilter } from "./FilterConverter";
import {
    toDateDataSetQualifier,
    toFactQualifier,
    toDisplayFormQualifier,
    toLocalIdentifier,
} from "../ObjRefConverter";
import compact = require("lodash/compact");
import get = require("lodash/get");

export function convertMeasure(measure: IMeasure): ExecuteAFM.IMeasure {
    const {
        measure: { definition },
    } = measure;

    const convertedDefinition = convertMeasureDefinition(definition);

    const format = getFormat(measure);
    const formatProp = format ? { format } : {};

    const alias = measure.measure.alias ? measure.measure.alias : measure.measure.title;
    const aliasProp = alias ? { alias } : {};

    return {
        localIdentifier: measure.measure.localIdentifier,
        definition: convertedDefinition,
        ...aliasProp,
        ...formatProp,
    };
}

function convertMeasureDefinition(definition: IMeasureDefinitionType): ExecuteAFM.MeasureDefinition {
    if (isMeasureDefinition(definition)) {
        return convertSimpleMeasureDefinition(definition);
    } else if (isPoPMeasureDefinition(definition)) {
        return convertPopMeasureDefinition(definition);
    } else if (isPreviousPeriodMeasureDefinition(definition)) {
        return convertPreviousPeriodMeasureDefinition(definition);
    } else if (isArithmeticMeasureDefinition(definition)) {
        return convertArithmeticMeasureDefinition(definition);
    } else {
        throw Error("The measure definition is not supported: " + JSON.stringify(definition));
    }
}

function convertAggregation(
    aggregation?: MeasureAggregation,
): ExecuteAFM.SimpleMeasureAggregation | undefined {
    if (!aggregation) {
        return undefined;
    }
    if (aggregation === "sum") {
        return "SUM";
    }
    if (aggregation === "avg") {
        return "AVG";
    }
    if (aggregation === "count") {
        return "COUNT";
    }
    if (aggregation === "max") {
        return "MAX";
    }
    if (aggregation === "median") {
        return "MEDIAN";
    }
    if (aggregation === "min") {
        return "MIN";
    }

    return "RUNSUM";
}

function convertSimpleMeasureDefinition(definition: IMeasureDefinition): ExecuteAFM.ISimpleMeasureDefinition {
    const { measureDefinition } = definition;

    const filters: ExecuteAFM.FilterItem[] = measureDefinition.filters
        ? compact(measureDefinition.filters.map(convertVisualizationObjectFilter))
        : [];
    const filtersProp = filters.length ? { filters } : {};

    const aggregation = convertAggregation(measureDefinition.aggregation);
    const aggregationProp = aggregation ? { aggregation } : {};

    const computeRatio = measureDefinition.computeRatio;
    const computeRatioProp = computeRatio ? { computeRatio } : {};

    const measureRef = measureDefinition.item;

    return {
        measure: {
            item: toFactQualifier(measureRef),
            ...filtersProp,
            ...aggregationProp,
            ...computeRatioProp,
        },
    };
}

function convertPopMeasureDefinition(
    definition: IPoPMeasureDefinition,
): ExecuteAFM.IOverPeriodMeasureDefinition {
    const { popMeasureDefinition } = definition;
    const attributeRef = popMeasureDefinition.popAttribute;

    return {
        overPeriodMeasure: {
            measureIdentifier: toLocalIdentifier(popMeasureDefinition.measureIdentifier),
            dateAttributes: [
                {
                    attribute: toDisplayFormQualifier(attributeRef),
                    periodsAgo: 1,
                },
            ],
        },
    };
}

function convertPreviousPeriodMeasureDefinition(
    definition: IPreviousPeriodMeasureDefinition,
): ExecuteAFM.IPreviousPeriodMeasureDefinition {
    const { previousPeriodMeasure } = definition;

    return {
        previousPeriodMeasure: {
            measureIdentifier: toLocalIdentifier(previousPeriodMeasure.measureIdentifier),
            dateDatasets: previousPeriodMeasure.dateDataSets.map((dateDataSet) => {
                const datasetRef = dateDataSet.dataSet;

                return {
                    dataset: toDateDataSetQualifier(datasetRef),
                    periodsAgo: dateDataSet.periodsAgo,
                };
            }),
        },
    };
}

function convertArithmeticMeasureDefinition(
    definition: IArithmeticMeasureDefinition,
): ExecuteAFM.IArithmeticMeasureDefinition {
    const { arithmeticMeasure } = definition;
    return {
        arithmeticMeasure: {
            measureIdentifiers: arithmeticMeasure.measureIdentifiers.map(toLocalIdentifier),
            operator: arithmeticMeasure.operator,
        },
    };
}

function getFormat(measure: IMeasure): string | undefined {
    const {
        measure: { definition },
    } = measure;
    const measureFormat = get(measure.measure, "format");

    if (isArithmeticMeasureDefinition(definition)) {
        if (definition.arithmeticMeasure.operator === "change") {
            return "#,##0.00%";
        }
    }

    const predefinedFormat = isMeasureDefinition(definition) ? getPredefinedFormat(definition) : undefined;

    return predefinedFormat || measureFormat;
}

function getPredefinedFormat(definition: IMeasureDefinition): string | null {
    const { measureDefinition } = definition;
    // should we prefer format defined on measure? If so, fix computeRatio format in AD
    return measureDefinition.computeRatio
        ? "#,##0.00%"
        : measureDefinition.aggregation === "count"
        ? "#,##0"
        : null;
}
