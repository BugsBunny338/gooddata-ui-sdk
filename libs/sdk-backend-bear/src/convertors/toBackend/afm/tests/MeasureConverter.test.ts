// (C) 2020 GoodData Corporation

import { convertMeasure } from "../MeasureConverter";
import { invalidMeasureDefinition } from "./InvalidInputs.fixture";
import { ReferenceLdm } from "@gooddata/reference-workspace";
import {
    newArithmeticMeasure,
    newPopMeasure,
    newPreviousPeriodMeasure,
    newMeasure,
    newPositiveAttributeFilter,
} from "@gooddata/sdk-model";

describe("measure converter", () => {
    const Scenarios: Array<[string, any]> = [
        ["local indentifier of measure", ReferenceLdm.Amount],
        [
            "converted arithmetic measure definition from model to AFM",
            newArithmeticMeasure(["foo", "bar"], "sum"),
        ],
        ["converted pop measure definition from model to AFM", newPopMeasure(ReferenceLdm.Won, "attr")],
        [
            "converted previous period measure from model to AFM",
            newPreviousPeriodMeasure("foo", [{ dataSet: "bar", periodsAgo: 3 }]),
        ],
        ["format of measure: change", newArithmeticMeasure(["foo", "bar"], "change")],
        ["format of measure: ratio", newMeasure("foo", m => m.ratio())],
        ["format of measure: count", newMeasure("foo", m => m.aggregation("count"))],
        [
            "converted definition",
            newMeasure("foo", m => m.filters(newPositiveAttributeFilter("filter", { uris: ["baz"] }))),
        ],
        ["converted alias", newMeasure("foo", m => m.alias("alias"))],
    ];
    it.each(Scenarios)("should return %s", (_disc, input) => {
        expect(convertMeasure(input)).toMatchSnapshot();
    });

    it("should throw an error when measure definition is not supported", () => {
        expect(() => convertMeasure(invalidMeasureDefinition)).toThrowErrorMatchingSnapshot();
    });
});
