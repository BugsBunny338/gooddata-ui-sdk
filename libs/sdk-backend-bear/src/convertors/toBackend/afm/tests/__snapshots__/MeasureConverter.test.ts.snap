// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`measure converter should return converted alias 1`] = `
Object {
  "alias": "alias",
  "definition": Object {
    "measure": Object {
      "item": Object {
        "identifier": "foo",
      },
    },
  },
  "localIdentifier": "m_627f0f0a_foo",
}
`;

exports[`measure converter should return converted arithmetic measure definition from model to AFM 1`] = `
Object {
  "definition": Object {
    "arithmeticMeasure": Object {
      "measureIdentifiers": Array [
        "foo",
        "bar",
      ],
      "operator": "sum",
    },
  },
  "localIdentifier": "m_3858f62230ac3c915f300c664312c63f",
}
`;

exports[`measure converter should return converted definition 1`] = `
Object {
  "definition": Object {
    "measure": Object {
      "filters": Array [
        Object {
          "positiveAttributeFilter": Object {
            "displayForm": Object {
              "identifier": "filter",
            },
            "in": Object {
              "uris": Array [
                "baz",
              ],
            },
          },
        },
      ],
      "item": Object {
        "identifier": "foo",
      },
    },
  },
  "localIdentifier": "m_foo_e2e8b1b1",
}
`;

exports[`measure converter should return converted pop measure definition from model to AFM 1`] = `
Object {
  "definition": Object {
    "popMeasure": Object {
      "measureIdentifier": "m_acugFHNJgsBy",
      "popAttribute": Object {
        "identifier": "attr",
      },
    },
  },
  "localIdentifier": "m_m_acugFHNJgsBy_attr",
}
`;

exports[`measure converter should return converted previous period measure from model to AFM 1`] = `
Object {
  "definition": Object {
    "previousPeriodMeasure": Object {
      "dateDataSets": Array [
        Object {
          "dataSet": Object {
            "identifier": "bar",
          },
          "periodsAgo": 3,
        },
      ],
      "measureIdentifier": "foo",
    },
  },
  "localIdentifier": "m_foo_previous_period",
}
`;

exports[`measure converter should return custom format: arithmetic measure 1`] = `
Object {
  "definition": Object {
    "arithmeticMeasure": Object {
      "measureIdentifiers": Array [
        "foo",
        "bar",
      ],
      "operator": "change",
    },
  },
  "format": "custom #,##0,00",
  "localIdentifier": "m_e70a9702_3858f62230ac3c915f300c664312c63f",
}
`;

exports[`measure converter should return custom format: pop measure 1`] = `
Object {
  "definition": Object {
    "popMeasure": Object {
      "measureIdentifier": "m_acugFHNJgsBy",
      "popAttribute": Object {
        "identifier": "attr",
      },
    },
  },
  "format": "custom #,##0,00",
  "localIdentifier": "m_e70a9702_m_acugFHNJgsBy_attr",
}
`;

exports[`measure converter should return custom format: previous period measure 1`] = `
Object {
  "definition": Object {
    "previousPeriodMeasure": Object {
      "dateDataSets": Array [
        Object {
          "dataSet": Object {
            "identifier": "bar",
          },
          "periodsAgo": 3,
        },
      ],
      "measureIdentifier": "foo",
    },
  },
  "format": "custom #,##0,00",
  "localIdentifier": "m_e70a9702_foo_previous_period",
}
`;

exports[`measure converter should return custom format: simple measure 1`] = `
Object {
  "definition": Object {
    "measure": Object {
      "item": Object {
        "identifier": "foo",
      },
    },
  },
  "format": "custom #,##0,00",
  "localIdentifier": "m_e70a9702_foo",
}
`;

exports[`measure converter should return format of measure: change 1`] = `
Object {
  "definition": Object {
    "arithmeticMeasure": Object {
      "measureIdentifiers": Array [
        "foo",
        "bar",
      ],
      "operator": "change",
    },
  },
  "format": "#,##0.00%",
  "localIdentifier": "m_3858f62230ac3c915f300c664312c63f",
}
`;

exports[`measure converter should return format of measure: count 1`] = `
Object {
  "definition": Object {
    "measure": Object {
      "aggregation": "count",
      "item": Object {
        "identifier": "foo",
      },
    },
  },
  "format": "#,##0",
  "localIdentifier": "m_foo_count",
}
`;

exports[`measure converter should return format of measure: ratio 1`] = `
Object {
  "definition": Object {
    "measure": Object {
      "computeRatio": true,
      "item": Object {
        "identifier": "foo",
      },
    },
  },
  "format": "#,##0.00%",
  "localIdentifier": "m_foo_ratio",
}
`;

exports[`measure converter should return local identifier of measure 1`] = `
Object {
  "definition": Object {
    "measure": Object {
      "item": Object {
        "identifier": "aangOxLSeztu",
      },
    },
  },
  "localIdentifier": "m_aangOxLSeztu",
}
`;

exports[`measure converter should throw an error when measure definition is not supported 1`] = `"The measure definition is not supported: {}"`;
